examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+
+
+"""
+## Complete Transformer Model
+
+Our model takes audio spectrograms as input, and predicts a sequence of characters.
+During training, we give the decoder the target character sequence shifted to the left
+as input. During inference, the decoder uses its own past predictions to predict the 
+next token.
+"""
+
+loss_object = tf.keras.losses.CategoricalCrossentropy(
+    from_logits=True, label_smoothing=0.1, reduction="none"
+)
+
+
+def masked_loss(real, pred):
+    """ assuming pad token index = 0 """
+    one_hot = tf.one_hot(real, depth=34)
+    mask = tf.math.logical_not(tf.math.equal(real, 0))
+    loss_ = loss_object(one_hot, pred)
+    mask = tf.cast(mask, dtype=loss_.dtype)
+    loss_ *= mask
+    return tf.reduce_mean(loss_)
+
+
+class Transformer(keras.Model):
+    def __init__(
+        self,
+        num_hid=64,
+        num_head=2,
+        num_ff=128,
+        src_maxlen=100,
+        trg_maxlen=100,
+        num_layers_enc=4,
+        num_layers_dec=1,
+        num_classes=10,
+    ):
+        super().__init__()
+        self.loss_metric = keras.metrics.Mean(name="loss")
+        self.num_layers_enc = num_layers_enc
+        self.num_layers_dec = num_layers_dec
+        self.trg_maxlen = trg_maxlen
+
+        self.enc_input = TransformerInput(
+            input_type="features", num_hid=num_hid, maxlen=src_maxlen
+        )
+        self.dec_input = TransformerInput(
+            input_type="tokens",
+            num_vocab=num_classes,
+            num_hid=num_hid,
+            maxlen=trg_maxlen,
+        )
+        for i in range(num_layers_enc):
+            setattr(
+                self,
+                f"enc_layer_{i}",
+                TransformerEncoderLayer(num_hid, num_head, num_ff),
+            )
+        for i in range(num_layers_dec):
+            setattr(
+                self,
+                f"dec_layer_{i}",
+                TransformerDecoderLayer(num_hid, num_head, num_ff),
+            )
+
+        self.final = layers.Dense(num_classes)
+
+    def encode_src(self, src):
+        x = self.enc_input(src)
+        for i in range(self.num_layers_enc):
+            x = getattr(self, f"enc_layer_{i}")(x)
+        return x
+
+    def decode(self, enc_out, trg):
+        y = self.dec_input(trg)
+        for i in range(self.num_layers_dec):
+            y = getattr(self, f"dec_layer_{i}")(enc_out, y)
+        return y
+
+    def call(self, inputs):
+        src = inputs[0]
+        trg = inputs[1]
+        x = self.encode_src(src)
+        y = self.decode(x, trg)
+        return self.final(y)
+
+    @property
+    def metrics(self):
+        return [self.loss_metric]
+
+    def train_step(self, batch):
+        """ Process one batch inside model.fit() """
+        src = batch["src"]
+        trg = batch["trg"]
+        dec_inp = trg[:, :-1]
+        dec_trg = trg[:, 1:]
+        with tf.GradientTape() as tape:
+            preds = self([src, dec_inp])
+            loss = masked_loss(dec_trg, preds)
```

Comment:
In general the inner code of a model should not rely on variables that aren't attributes of the model (for the model to be portable it needs to be self-contained). Here the loss should be configured in compile, so that you do e.g. `self.compiled_loss(y_true, y_pred, sample_weight=mask)`

or you could pass the loss in the constructor and set it as an attr.